types/wlr_output_layout.c

```diff
@@ -190,6 +229,39 @@ static struct wlr_output_layout_output *output_layout_output_create(
 	return l_output;
 }
 
+void relocate_output_with_children(struct wlr_output_layout *layout,
+		struct wlr_output_layout_output *output,
+		struct wl_list *ref) {
+
+	struct wlr_output_layout_output *child_output;
+	struct wlr_output_layout_output *parent_output =
+		output;
+
+	if (output->link.prev == ref) {
+		return;
+	}
+
+	for (child_output = wl_container_of((output)->link.next, output, link);
+			&output->link != &layout->outputs;
+			child_output =
+				wl_container_of((child_output)->link.next, output, link)) {
+		while (child_output->reference != parent_output) {
+			if (parent_output == output) {
+				goto move_outputs;
+			}
+			parent_output = parent_output->reference;
+		}
+	}
+
+move_outputs:
+	ref->next->prev = child_output->link.prev; // A->prev = B
```

Comment:
Ah, so that's the code ordering outputs in topological ordering. I bet you can rewrite those manual pointer manipulations with `wl_list_remove` and `wl_list_insert`.

Reply:
Yes I can, I was just being too perfectionistic w.r.t. the number of required operations.